### PR TITLE
(122) Finish ingest of files - API Changes

### PR DESCRIPTION
**Pull request by esteedqueen**
1) Update SubmissionFile with number of rows to be ingested
    -- Add rows column to SubmissionFile
    -- Modify PATCH v1/submissions/:submission_id/files/:id endpoint to update rows column and add tests
   -- Modify POST and PATCH endpoints for submission_file_controller and tests to json api format

2) API endpoint `PATCH /v1/files/:file_id/entries/:entry_id` stores validation status and errors received from DAVE

3) Add `PATCH /v1/tasks/:task_id` endpoint for updating a task's attributes

4) Add `Add `GET v1/files/:id` endpoint to retrieve submission_file details and add validation_errors to submission_entry json response

### CODE REVIEW

File: spec/requests/v1/submissions_spec.rb [hunk 1]
```diff
@@ -80,7 +80,20 @@
     it 'creates a new submission file and returns its id' do
       submission = FactoryBot.create(:submission)
 
-      post "/v1/submissions/#{submission.id}/files"
+      headers = {
+        'Content-Type': 'application/vnd.api+json',
+        'Accept': 'application/vnd.api+json'
+      }
+
+      params = {
```
Comment (leeky) [inline spec/requests/v1/submissions_spec.rb]: FYI you can remove this as the attributes hash is empty.

File: app/controllers/v1/submission_files_controller.rb [hunk 2]
```diff
@@ -1,6 +1,7 @@
 class V1::SubmissionFilesController < ApplicationController
+  deserializable_resource :submission_file, only: %i[update]
   def create
-    submission = Submission.find(submission_file_params[:submission_id])
+    submission = Submission.find(params[:submission_id])
```
Comment (leeky) [inline app/controllers/v1/submission_files_controller.rb]: :+1:

File: app/controllers/v1/submission_entries_controller.rb [hunk 3]
```diff
@@ -16,9 +16,15 @@ def create
 
   def update
     submission_file = SubmissionFile.find(params[:file_id])
-    submission_file.entries.find(params[:id])
+    entry = submission_file.entries.find(params[:id])
+    entry.aasm.current_state = submission_entry_params[:status]
```
Comment (leeky) [inline app/controllers/v1/submission_entries_controller.rb]: Have you considered checking if `submission_entry_params[:status]` is empty? If it's omitted, would this set the entry's status to `''` ?

File: app/controllers/v1/submission_files_controller.rb [hunk 4]
```diff
@@ -13,7 +13,7 @@ def create
 
   def update
     submission_file = SubmissionFile.find(params[:id])
-    submission_file.rows = submission_file_params[:rows]
+    submission_file.rows = submission_file_params[:rows] if submission_file_params[:rows]
```
Comment (leeky) [inline app/controllers/v1/submission_files_controller.rb]: How about using a guard clause? If there are no rows to update, we have nothing to do.

File: spec/requests/v1/files_spec.rb [hunk 5]
```diff
@@ -103,7 +103,8 @@
         data: {
           type: 'submission_entries',
           attributes: {
-            valid: true
+            status: 'errored',
+            validation_errors: { message: 'Invalid data' }
```
Comment (leeky) [inline spec/requests/v1/files_spec.rb]: Maybe a more realistic `validation_errors` payload would be useful here?

File: spec/requests/v1/files_spec.rb [hunk 6]
```diff
@@ -116,6 +117,8 @@
       patch "/v1/files/#{file.id}/entries/#{entry.id}", params: params.to_json, headers: headers
 
       expect(response).to have_http_status(:no_content)
+      expect(entry.reload.aasm.current_state).to eql :errored
```
Comment (leeky) [inline spec/requests/v1/files_spec.rb]: `ActsAsStateMachine` adds some query methods to the model, e.g. `errored?`, so this would also work:

`expect(entry.reload).to be_errored`

File: app/controllers/v1/tasks_controller.rb [hunk 7]
```diff
@@ -25,6 +25,17 @@ def show
     render jsonapi: task, include: params.dig(:include)
   end
 
+  def update
+    task = Task.find(params[:id])
+    task.update(task_params)
+
+    if task.save
+      render jsonapi: task
```
Comment (leeky) [inline app/controllers/v1/tasks_controller.rb]: As this is a `PATCH`, it shouldn't return any content.

File: spec/requests/v1/tasks_spec.rb [hunk 8]
```diff
@@ -162,4 +162,32 @@
...
+        'Accept': 'application/vnd.api+json'
+      }
+
+      patch "/v1/tasks/#{task.id}", params: params.to_json, headers: headers
+
+      expect(response).to be_successful
+
+      task.reload
+
+      expect(task).to be_in_progress
```
Comment (leeky) [inline spec/requests/v1/tasks_spec.rb]: :+1:

File: spec/requests/v1/submissions_spec.rb [hunk 9]
```diff
@@ -130,6 +130,23 @@
...
+      submission = FactoryBot.create(:submission)
+      file = FactoryBot.create(:submission_file, submission_id: submission.id, rows: 40)
+
+      get "/v1/submissions/#{submission.id}/files/#{file.id}"
+
+      expect(response).to have_http_status(:ok)
+
+      expect(json['data']).to have_id(file.id)
+
+      expect(json['data']).to have_attribute(:rows).with_value(file.rows)
```
Comment (leeky) [inline spec/requests/v1/submissions_spec.rb]: Don't this test and the one immediately below test the same thing?

File: app/serializable/serializable_submission_entry.rb [hunk 10]
```diff
@@ -7,6 +7,8 @@ class SerializableSubmissionEntry < JSONAPI::Serializable::Resource
   attribute :source
   attribute :data
 
+  attribute :validation_errors
```
Comment (leeky) [inline app/serializable/serializable_submission_entry.rb]: Unless I'm mistaken, this seems to be unrelated to the rest of the commit?

Comment (esteedqueen) [reply app/serializable/serializable_submission_entry.rb]: It is used during checks for validation in the frontend code.

File: app/controllers/v1/submission_entries_controller.rb (hunk 3 above)
Comment (esteedqueen) [reply app/controllers/v1/submission_entries_controller.rb]: Yeah I checked, I think aasm already checks that current_state can't be nil, so I thought checking again will be a bit redundant, but happy to add it.

Comment (leeky) [reply app/controllers/v1/submission_entries_controller.rb]: Happy to leave this as-is. Perhaps add a test around updating `validation_errors` and not `status` to verify that `aasm` doesn't blow up in this case?

File: app/serializable/serializable_submission_entry.rb (hunk 10 above)
Comment (leeky) [reply app/serializable/serializable_submission_entry.rb]: Fair enough :+1:

**Comment by esteedqueen** [comment]
@leeky  this is ready for a second look.

**Comment by leeky** [comment]
@esteedqueen Thanks - LGTM! :shipit:
